### PR TITLE
[SPIR-V] Rework usage of virtual registers' types and classes

### PR DESCRIPTION
**Pull request by VyacheslavLevytskyy**
This PR continues https://github.com/llvm/llvm-project/pull/101732 changes in virtual register processing aimed to improve correctness of emitted MIR between passes from the perspective of MachineVerifier. Namely, the following changes are introduced:
* register classes (lib/Target/SPIRV/SPIRVRegisterInfo.td) and instruction patterns (lib/Target/SPIRV/SPIRVInstrInfo.td) are corrected and simplified (by removing unnecessary sophisticated options) -- e.g., this PR gets rid of duplicating 32/64 bits patterns, removes ANYID register class and simplifies definition of the rest of register classes,
* hardcoded LLT scalar types in passes before instruction selection are corrected -- the goal is to have correct bit width before instruction selection, and use 64 bits registers for pattern matching in the instruction selection pass; 32-bit registers remain where they are described in such terms by SPIR-V specification (like, for example, creation of virtual registers for scope/mem semantics operands),
* rework virtual register type/class assignment for calls/builtins lowering,
* a series of minor changes to fix validity of emitted code between passes:
  - ensure that that bitcast changes the type,
  - fix the pattern for instruction selection for OpExtInst,
  - simplify inline asm operands usage,
  - account for arbitrary integer sizes / update legalizer rules;
* add '-verify-machineinstrs' to existed test cases.

See also https://github.com/llvm/llvm-project/issues/88129 that this PR may resolve.

This PR fixes a great number of issues reported by MachineVerifier and, as a result, reduces a number of failed test cases for the mode with expensive checks set on from ~200 to ~57.

### CODE REVIEW

**Review by michalpaszkowski — APPROVED** [review]
Thank you @VyacheslavLevytskyy, this is a huge change and a very great progress, I can only imagine the time it took. The change looks good to me from the code perspective. However, please wait with merging for a couple of hours. I am running a full OpenCL CTS testing cycle + benchmarks and it should finish around your afternoon (I will add one more comment).

File: llvm/lib/Target/SPIRV/SPIRVGlobalRegistry.cpp [hunk 1]
```diff
@@ -407,7 +393,7 @@ Register SPIRVGlobalRegistry::getOrCreateCompositeOrNull(
 
     // TODO: handle cases where the type is not 32bit wide
```
Comment (farzonl) [inline llvm/lib/Target/SPIRV/SPIRVGlobalRegistry.cpp:394]: todo doesn't seem to make sense anymore.

File: llvm/lib/Target/SPIRV/SPIRVGlobalRegistry.cpp [hunk 2]
```diff
@@ -1451,3 +1438,61 @@ Register SPIRVGlobalRegistry::getOrCreateUndef(MachineInstr &I,
...
+    SPIRVType *ElemType = getSPIRVTypeForVReg(SpvType->getOperand(1).getReg());
+    unsigned ElemOpcode = ElemType ? ElemType->getOpcode() : 0;
+    if (ElemOpcode == SPIRV::OpTypeFloat)
+      return &SPIRV::vfIDRegClass;
+    if (ElemOpcode == SPIRV::OpTypePointer)
+      return &SPIRV::vpIDRegClass;
+    return &SPIRV::vIDRegClass;
+  }
+  }
+  return &SPIRV::iIDRegClass;
```
Comment (farzonl) [inline llvm/lib/Target/SPIRV/SPIRVGlobalRegistry.cpp:1460]: maybe add an assert for `SPIRV::OpTypeInt`?

**Review by farzonl — APPROVED** [review]

**Comment by farzonl** [comment]
LGTM! Thank you for updating the HLSL tests!

**Comment by michalpaszkowski** [comment]
> Thank you @VyacheslavLevytskyy, this is a huge change and a very great progress, I can only imagine the time it took. The change looks good to me from the code perspective. However, please wait with merging for a couple of hours. I am running a full OpenCL CTS testing cycle + benchmarks and it should finish around your afternoon (I will add one more comment).

Confirming there are no regressions with OpenCL CTS

**Pull request merged**